ResponsesAgent docs (#15741)
mlflow/mlflow: docs/docs/llms/responses-agent-intro/index.mdx

import { APILink } from "@site/src/components/APILink";

# Tutorial: ResponsesAgent Introduction

## What is a ResponsesAgent?

> `ResponsesAgent` is a subclass of `PythonModel` that provides a framework-agnostic way to create an agent model. Authoring a `ResponsesAgent` provides the following benefits:

Reviewer: ```suggestion
`ResponsesAgent` is a subclass of `PythonModel` that provides a framework-agnostic way to create an agent model. Authoring an agent using `ResponsesAgent` provides the following benefits:
```

> - Ensure your model is compatible with OpenAI Responses API (Read more about it in the [Responses API Introduction](./responses-api-intro))

Reviewer: I would emphasize the other bullets first - I think the benefit to call out here is that you can author an agent that is compatible with standard clients like the OAI responses client, simplifying writing downstream UIs or applications

> - Support for multiple messages and intermediate steps

Reviewer: ```suggestion
- Support for returning multiple output messages, including intermediate outputs from tool-calling
```

- Support for multi-agent scenarios
- Ensure compatibility with MLflow logging, tracing, and model serving

> We recommend `ResponsesAgent` instead of <APILink fn="mlflow.pyfunc.ChatModel">`ChatModel`</APILink> and <APILink fn="mlflow.pyfunc.ChatAgent">`ChatAgent`</APILink>, as it has all the benefits of `ChatAgent` and is compatible with OpenAIs Responses API.

Reviewer: I guess we should update these pages also to note that Responses is preferred

Reviewer: Oh nice looks like we already did?

## Author a ResponsesAgent

> To author a ResponsesAgent, create a subclass of the ResponsesAgent and implement a predict method using any agent authoring framework. You must be using `pydantic>=2` to use ResponsesAgent.

Reviewer: ```suggestion
To author a ResponsesAgent, create a subclass of `mlflow.pyfunc.ResponsesAgent` and implement your agent logic in its `predict` method. You can use any agent authoring framework(s) in your implementation. You must be using `pydantic>=2` to use ResponsesAgent.
```

> When creating output objects, you should only create `ResponsesAgentResponse` and `ResponsesAgentStreamEvent` pydantic objects. The other classes in `mlflow.types.responses_helpers` are for validation only. Below is an example of a tool calling sequence:

Reviewer: Feels slightly low level / too detailed for something you see when first reading about the interface?

Reviewer: Also maybe worth mentioning the streaming stuff later on - and is it worth maybe having a "Getting started" section with this content + then following up with "Streaming" as a more advanced capability?

                    "type": "output_text",
                }
            ],
            "role": "assistant",
            "status": "completed",
            "type": "message",
        },
    ]
)
```
An equivalent stream would look like:
```python
from mlflow.types.responses import ResponsesAgentStreamEvent

ResponsesAgentStreamEvent(
    type="response.output_item.done",
    item={
        "arguments": '{"code":"result = 4 * 3\\nprint(result)"}',
        "call_id": "call_1",
        "name": "python_exec",
        "type": "function_call",
        "id": "fc_1",
        "status": "completed",
    },
)
ResponsesAgentStreamEvent(
    type="response.output_item.done",
    item={
        "type": "function_call_output",
        "call_id": "call_1",
        "output": "12\n",
    },
)
ResponsesAgentStreamEvent(
    type="response.output_item.done",
    item={
        "id": "msg_1",
        "content": [
            {
                "annotations": [],
                "text": "The result of 4 * 3 in Python is 12.",
                "type": "output_text",
            }
        ],
        "role": "assistant",
        "status": "completed",
        "type": "message",
    },
)
```

> ### Streaming a ResponsesAgent

Reviewer: ```suggestion
### Streaming agent output
```

Would be good to focus on the end goal vs the specific class/interface name

See the [example below](#example-responsesagent) for a full implementation!

> 

Reviewer: This seems like maybe the right place to mention the details about streaming classes etc

> ## Log a ResponsesAgent

Reviewer: ```suggestion
## Log your agent
```

> Log your ResponsesAgent using the [Models-from-code](/model/models-from-code) approach. This approach is framework-agnostic and supports all authoring frameworks:

Reviewer: ```suggestion
Log your agent using the [Models-from-code](/model/models-from-code) approach. This approach is framework-agnostic and supports all authoring frameworks:
```

- Metadata
    - `{"task": "agent/v1/responses"}` will be automatically appended to any metadata that you may pass in when logging the model
- Input Example
    - Providing an input example is optional, `mlflow.types.responses.RESPONSES_AGENT_INPUT_EXAMPLE` will be used by default
    - If you do provide an input example, ensure it's a dictionary of the ResponsesAgentRequest schema

## Testing out a ResponsesAgent
To test out a ResponsesAgent, you can pass a single input dictionary that follows the ResponsesAgentRequest schema both before and after logging it:
```python
from mlflow.pyfunc import ResponsesAgent
class MyResponsesAgent(ResponsesAgent):
    ...
responses_agent = MyResponsesAgent()
responses_agent.predict(
    {
        "input": [{"role": "user", "content": "what is 4*3 in python"}],
        "context": {"conversation_id": "123", "user_id": "456"},
    }
)
# log responses_agent
loaded_model = mlflow.pyfunc.load_model(path)
loaded_model.predict(
    {
        "input": [{"role": "user", "content": "what is 4*3 in python"}],
        "context": {"conversation_id": "123", "user_id": "456"},
    }
)
```

> ## Example ResponsesAgent

Reviewer: ```suggestion
## Example
```

> Here's an example of a ResponsesAgent that calls OpenAI's gpt-4o model with a simple tool.

Reviewer: ```suggestion
Here's an example agent that calls OpenAI's gpt-4o model with a simple tool.
```

import { APILink } from "@site/src/components/APILink";

# ResponsesAgent Introduction

## What is a ResponsesAgent?
`ResponsesAgent` is a subclass of `PythonModel` that provides a framework-agnostic way to create an agent model. Authoring an agent using `ResponsesAgent` provides the following benefits:
- Support for returning multiple output messages, including intermediate outputs from tool-calling
- Support for multi-agent scenarios
- Ensure compatibility with MLflow logging, tracing, and model serving
- Ensure your model is compatible with OpenAI Responses API, making it compatible with OpenAI's responses client and other downstream UIs/applications

> We recommend `ResponsesAgent` instead of <APILink fn="mlflow.pyfunc.ChatModel">`ChatModel`</APILink> and <APILink fn="mlflow.pyfunc.ChatAgent">`ChatAgent`</APILink>, as it has all the benefits of `ChatAgent` and and supports additional features like annotations.

Reviewer 2: Does this mean we should deprecate ChatModel & ChatAgent?

Reviewer 3: This is my understanding, @bbqiu can we show deprecation note in these chat model guides as well? https://output.circle-artifacts.com/output/job/95c42e89-8729-440f-b31b-10cd0c2548ad/artifacts/0/docs/build/latest/llms/chat-model-intro/

Author: as noted from above, will not mark ChatAgent as deprecated just yet: https://github.com/mlflow/mlflow/pull/15741#discussion_r2094733627

will mark chatmodel as deprecated

> We recommend `ResponsesAgent` instead of <APILink fn="mlflow.pyfunc.ChatModel">`ChatModel`</APILink> and <APILink fn="mlflow.pyfunc.ChatAgent">`ChatAgent`</APILink>, as it has all the benefits of `ChatAgent` and and supports additional features like annotations.

Author: will do this in a followup in that integrated "authoring an agent" guide

## Author a ResponsesAgent

### Getting started
To create your own agent, subclass `mlflow.pyfunc.ResponsesAgent` and implement your agent logic in the `predict` method. The implementation is framework-agnostic, allowing you to use any agent authoring framework. Note that `pydantic>=2` is required to use ResponsesAgent. For a complete implementation example, see the [example below](#example).

> 

Reviewer 3: Can we start with the simplest possible example here? For example, just calls OpenAI Responses API and return it back, or even dummy code without actual predict implementation?

The tool calling example is too complex to get started. We talk about interface and data models next, but it's easier if users see the simple method signature here before digging into details.

Author: added a simple chat example

### Creating agent output
When implementing your agent, you'll work with two main output types: `ResponsesAgentResponse` and `ResponsesAgentStreamEvent`. These are the only pydantic objects you should create directly. The remaining classes in `mlflow.types.responses_helpers` are only for validating dictionaries.

> Below are some helper methods you can use to create common output item types:

Reviewer 2: Are those commonly used? If yes should we add them as helper functions?

Author: will add them -- didn't think i was allowed to but i saw the msg in mlflow3 about EOD 5/20 instead

> Below are some helper methods you can use to create common output item types:

Author: clarified w/ example usage,

)
ResponsesAgentStreamEvent(
    type="response.output_item.done",
    item={
        "type": "function_call_output",
        "call_id": "call_1",
        "output": "12\n",
    },
)
ResponsesAgentStreamEvent(
    type="response.output_item.done",
    item={
        "id": "msg_1",
        "content": [
            {
                "annotations": [],
                "text": "The result of 4 * 3 in Python is 12.",
                "type": "output_text",
            }
        ],
        "role": "assistant",
        "status": "completed",
        "type": "message",
    },
)
```

## Log your agent

Log your agent using the [Models-from-code](/model/models-from-code) approach. This approach is framework-agnostic and supports all authoring frameworks:
```python
with mlflow.start_run():
    logged_agent_info = mlflow.pyfunc.log_model(
        artifact_path="agent",
        python_model="agent.py",  # replace with your relative path to agent code
    )
```

For ease of use, MLflow has built in the following features:
- Automatic model signature inference
    - An input and output signature that adheres to the ResponsesAgentRequest and ResponsesAgentResponse schemas will be set
- Metadata

>     - `{"task": "agent/v1/responses"}` will be automatically appended to any metadata that you may pass in when logging the model

Reviewer 2: Could you explain what's the usage of this field? Why do users care?

Author: users wouldn't really care, but I wouldn't want users to be surprised by additional metadata being logged (though i'm not sure if they would)

databricks uses this to enable a bunch of serving features (streaming, gateway payload logging etc.) but i didn't want to mention it in the OSS docs

- Input Example
    - Providing an input example is optional, `mlflow.types.responses.RESPONSES_AGENT_INPUT_EXAMPLE` will be used by default
    - If you do provide an input example, ensure it's a dictionary of the ResponsesAgentRequest schema

## Testing out your agent
To test out a ResponsesAgent, you can pass a single input dictionary that follows the ResponsesAgentRequest schema both before and after logging it:
```python
from mlflow.pyfunc import ResponsesAgent

class MyResponsesAgent(ResponsesAgent):
    ...

responses_agent = MyResponsesAgent()
responses_agent.predict(
    {
        "input": [{"role": "user", "content": "what is 4*3 in python"}],
        "context": {"conversation_id": "123", "user_id": "456"},
    }
)
# log the responses_agent
loaded_model = mlflow.pyfunc.load_model(path)
loaded_model.predict(
    {
        "input": [{"role": "user", "content": "what is 4*3 in python"}],
        "context": {"conversation_id": "123", "user_id": "456"},
    }
)
```

## Example
Here's an example of an agent that calls OpenAI's gpt-4o model with a simple tool.

```python

> # if using a jupyter notebook

Reviewer 2: ```suggestion
# uncomment below if running inside a jupyter notebook
```

# %%writefile agent.py
import json
from typing import Any, Callable, Generator
from uuid import uuid4

import backoff
import mlflow
import openai
from mlflow.entities import SpanType
from mlflow.pyfunc import ResponsesAgent
from mlflow.types.responses import (
    ResponsesAgentRequest,
    ResponsesAgentResponse,
    ResponsesAgentStreamEvent,
)
from openai import OpenAI
from pydantic import BaseModel

class ToolInfo(BaseModel):
    """
    Class representing a tool for the agent.
    - "name" (str): The name of the tool.
    - "spec" (dict): JSON description of the tool (matches OpenAI Responses format)
    - "exec_fn" (Callable): Function that implements the tool logic
    """

    name: str
    spec: dict
    exec_fn: Callable

class ToolCallingAgent(ResponsesAgent):
    """
    Class representing a tool-calling Agent
    """

    def __init__(self, llm_endpoint: str, tools: list[ToolInfo]):
        """Initializes the ToolCallingAgent with tools."""

>         self.llm_endpoint = llm_endpoint

Reviewer 3: ```suggestion
        self.model = model
```

nit: let's follow the natural terminology of OAI SDK here

        )

    @mlflow.trace(span_type=SpanType.AGENT)
    def predict_stream(
        self, request: ResponsesAgentRequest
    ) -> Generator[ResponsesAgentStreamEvent, None, None]:
        input_messages = [{"role": "system", "content": SYSTEM_PROMPT}] + [
            i.model_dump() for i in request.input
        ]
        yield from self.call_and_run_tools(input_messages=input_messages)

tools = [
    ToolInfo(
        name="get_weather",
        spec={
            "type": "function",
            "name": "get_weather",
            "description": "Get current temperature for provided coordinates in celsius.",
            "parameters": {
                "type": "object",
                "properties": {
                    "latitude": {"type": "number"},
                    "longitude": {"type": "number"},
                },
                "required": ["latitude", "longitude"],
                "additionalProperties": False,
            },
            "strict": True,
        },
        exec_fn=lambda latitude, longitude: 70,  # dummy tool implementation
    )
]
SYSTEM_PROMPT = "You are a helpful assistant that can call tools to get information."

> mlflow.openai.autolog()

Reviewer 2: Do we want to mention users need to set OPENAI_API_KEY?

                        "type": "function",
                        "function": {
                            "arguments": message["arguments"],
                            "name": message["name"],
                        },
                    }
                ],
            }
        ]
    elif msg_type == "message" and isinstance(message["content"], list):
        return [
            {"role": message["role"], "content": content["text"]}
            for content in message["content"]
        ]
    elif msg_type == "function_call_output":
        return [
            {
                "role": "tool",
                "content": message["output"],
                "tool_call_id": message["call_id"],
            }
        ]
    compatible_keys = ["role", "content", "name", "tool_calls", "tool_call_id"]
    return [{k: v for k, v in message.items() if k in compatible_keys}]
```

The `ResponsesAgent` interface extends all functionality previously available in <APILink fn="mlflow.pyfunc.ChatAgent">`ChatAgent`</APILink>, while introducing new features. Below, we outline the key differences in message representation between the two interfaces for common use cases:

> ### Standard Text Response

Reviewer 2: Could we use `<table>` for this section for better comparison between two models?

Author: was unable to get multiline code examples working here, but will see if i can do this in a followup https://databricks.atlassian.net/browse/ML-53898

> ### Standard Text Response

Reviewer 3: This comparison is great. Instead of raw data model only, can we show an example of how the prediction query looks different in a side-by-side table format? Sth like this

```
| Usage | ChatAgent | ResponsesAgent |
|:--:|:--:|:--:|
| Invoke with messages| ...| ...|
| Streaming |...|...|
| Using custom inputs |...|...|
| Passing tools |...|...|
...
```

      "type": "function",
      "function": {
        "name": "",
        "arguments": ""
      }
    }
  ]
}
```

---

### Tool Call Results

#### ResponsesAgent

```json
{
  "type": "function_call_output",
  "call_id": "call_1",
  "output": ""
}
```

#### ChatAgent

```json
{
  "role": "tool",
  "content": "12",
  "tool_call_id": "call_1"
}
```

---

### Tool Definitions

#### ResponsesAgent

```json
{
  "name": "",
  "parameters": {},
  "strict": true,
  "type": "function",
  "description": ""
}
```

#### ChatAgent

```json
{
  "type": "function",
  "function": {
    "name": "",
    "description": "",
    "parameters": {},
    "strict": true
  }
}
```

## Simple Chat Example
Here's an example of an agent that calls OpenAI's gpt-4o model with a simple tool:

```python
# uncomment below if running inside a jupyter notebook
# %%writefile agent.py
import os
from typing import Generator

import mlflow
from mlflow.entities.span import SpanType
from mlflow.models import set_model
from mlflow.pyfunc.model import ResponsesAgent
from mlflow.types.responses import (
    ResponsesAgentRequest,
    ResponsesAgentResponse,
    ResponsesAgentStreamEvent,
)
from openai import OpenAI

> 

Author: added!

import { APILink } from "@site/src/components/APILink";

# ResponsesAgent Introduction

## What is a ResponsesAgent?
`ResponsesAgent` is a subclass of `PythonModel` that provides a framework-agnostic way to create an agent model. Authoring an agent using `ResponsesAgent` provides the following benefits:
- Support for returning multiple output messages, including intermediate outputs from tool-calling
- Support for multi-agent scenarios
- Ensure compatibility with MLflow logging, tracing, and model serving
- Ensure your model is compatible with OpenAI Responses API, making it compatible with OpenAI's responses client and other downstream UIs/applications

> We recommend `ResponsesAgent` instead of <APILink fn="mlflow.pyfunc.ChatModel">`ChatModel`</APILink> and <APILink fn="mlflow.pyfunc.ChatAgent">`ChatAgent`</APILink>, as it has all the benefits of `ChatAgent` and is compatible with OpenAI's Responses API.

Reviewer: ```suggestion
We recommend `ResponsesAgent` instead of <APILink fn="mlflow.pyfunc.ChatModel">`ChatModel`</APILink> and <APILink fn="mlflow.pyfunc.ChatAgent">`ChatAgent`</APILink>, as it has all the benefits of `ChatAgent` and supports additional features like annotations
```

Just trying to avoid being too aligned with any particular LLM vendor & call out some of the tangible benefits instead
